StreamChatCoreTests/Unit Tests/ClientLoggerTests.swift

```diff
@@ -0,0 +1,168 @@
+                        .ban(.init(user: testUser, channel: testChannel, timeoutInMinutes: 0, reason: nil))]
+    }
+    
+    func testLogEndpointStrings() {
+        for endpoint in allEndpoints {
+            logger.log("Endpoint: \(String(describing: endpoint))", level: .debug)
+        }
+        
+        XCTAssertEqual(logged.count, allEndpoints.count)
+    }
+    
+    func testLogURLRequest() {
+        for endpoint in allEndpoints {
+            var urlRequest = URLRequest(url: testUrl)
+            urlRequest.httpMethod = endpoint.method.rawValue
+            urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
+            
+            if let body = endpoint.body {
+                let encodable = AnyEncodable(body)
+                
+                do {
+                    if let httpBody = try? JSONEncoder.defaultGzip.encode(encodable) {
+                        urlRequest.httpBody = httpBody
+                        urlRequest.addValue("gzip", forHTTPHeaderField: "Content-Encoding")
+                    } else {
+                        urlRequest.httpBody = try JSONEncoder.default.encode(encodable)
+                    }
+                } catch {
+                    XCTFail(error.localizedDescription)
+                }
+            }
+            
+            logger.log(urlRequest)
+        }
+        
+        XCTAssertFalse(logged.isEmpty)
+    }
+    
+    func testLogHeaders() {
+        let headers = [
+            "X-Stream-Client": "stream-chat-swift-client-2",
+            "X-Stream-Device": "iPhone11",
+            "X-Stream-OS": "iOS13.3",
+            "Stream-Auth-Type": "anonymous"]
+        
+        logger.log(headers: headers)
+        
+        XCTAssertFalse(logged.isEmpty)
+        for (key, value) in headers {
+            XCTAssert(logged[0].contains("\(key) = \(value)"))
+        }
+    }
+    
+    func testLogQueryItems() {
+        let queryItems: [URLQueryItem] = [.init(name: "test1", value: "test1"),
+                                          .init(name: "test2", value: "test2")]
+        
+        logger.log(queryItems)
+        
+        XCTAssertFalse(logged.isEmpty)
+        XCTAssertEqual(logged[0], "URL query items:\n▫️ test1=test1\n▫️ test2=test2\n")
+    }
+    
+    func testLogURLResponse() {
+        let urlResponse = URLResponse(url: testUrl, mimeType: nil, expectedContentLength: 1, textEncodingName: nil)
+        
+        logger.log(urlResponse, data: testData)
+        
+        XCTAssertFalse(logged.isEmpty)
+        XCTAssert(logged[0].starts(with: "⬅️❔ Unknown response (6 bytes): "))
+    }
+    
+    func testLogError() {
+        let testError = NSError(domain: "test", code: 1, userInfo: nil)
+        logger.log(testError as Error, message: "test error message")
+        
+        XCTAssertFalse(logged.isEmpty)
+        XCTAssertEqual(logged[0], "❌ test error message Error Domain=test Code=1 \"(null)\" in testLogError()[144]")
```

Comment:
why is [144] part of the message?

Reply:
It's really helpful information of the source of the error: a function + a line of code.